woocommerce/woocommerce
# Add block template with abstract blocks and templates

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR makes a few changes to the proposed API in https://github.com/woocommerce/woocommerce/pull/39470.  @mattsherman did a good job of summarizing the debate in https://github.com/woocommerce/woocommerce/pull/39470#issuecomment-1663934782.  The primary concerns I had around the API in that branch were:

* `add_block` is a public property.  In the case of WC templates, we will often not want this exposed and I would prefer not to override this with a warning (as this feels like a workaround to the declared interface).
* The `creator` method adds some overhead complexity and may make debugging slightly more difficult.  It also allows consumers to bypass restrictions on blocks.  I would prefer that the factory (i.e. block or template class) directly handle logic for how blocks should be instantiated.

@mattsherman is AFK at the moment, but his final primary concern was around not exposing concrete classes due to maintenance concerns and template blocks becoming out of sync, both of which are valid concerns.  I believe this PR keeps a more config-based, factory approach with a few key differences:

* Abstractions have been made for `AbstractBlock` and `AbstractBlockTemplate` that can be used to create custom blocks/templates with the bare minimum interface.  The `Block` and `BlockTemplate` classes still exist for convenience, but these could potentially be removed depending on whether or not they serve a real benefit or not.
* The `add_block` method has been moved to a protected `add_inner_block` method.  Templates and blocks are responsible for creating their own public interfaces for handling addition of blocks.
* There are no longer 2 methods for `add_block` and `add_container_block`.  If a template wanted to implement its own methods for this or prevent certain block types, it can choose to do that, but the generic `Block` class has container traits and can have inner children.

```php
class MyCustomBlock {
    public function add_custom_inner_block( $args ) {
        $custom_block = new CustomBlock( $args );
        $this->add_inner_block( $custom_block );
    }
}

$my_custom_block = new MyCustomBlock();
$my_custom_block->add_custom_inner_block( ... ); // This custom inner block does not expose the add_block method and can restrict to certain types of blocks.

$generic_block = new Block();
$generic_block->add_block( ... ); // This is a generic block that exposes the add_block method and by design is not restrictive.
```

Closes # .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Verify that tests are passing
2. Check that the templates make sense for our use case



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
